### ISSUE TITLE
impl Clone for rusqlite::Connection?

### ISSUE THREAD

**Issue by mqudsi** [jgallagher/rusqlite#393]
So long as SQLite is not running in single-threaded mode (and the default is that it's not), it should be safe to offer a `.clone()` implementation that would re-open the same database file with same settings/configuration and return the `Sync` result, while playing nicely with `!Sync`.

**Comment by thomcc** [comment]
This seems difficult if the database was in-memory, and probably requires tracking which pragmas were executed.

**Comment by gwenn** [comment]
If the in-memory mode is specified using [URI filenames](https://sqlite.org/uri.html)/[also](https://sqlite.org/c3ref/open.html#urifilenamesinsqlite3open), it seems feasible.

**Comment by thomcc** [comment]
Right, that's what I meant by 'a named in-memory database'. There's not really any guarantee that's true.

**Comment by gwenn** [comment]
If [`Connection.path.is_some()`](https://docs.rs/rusqlite/0.14.0/src/rusqlite/lib.rs.html#199) and if we keep the open flags, we can clone. 
But we can't use the standard trait because [clone](https://doc.rust-lang.org/std/clone/trait.Clone.html#tymethod.clone) cannot fail.

Why not use [r2d2](https://docs.rs/r2d2_sqlite/0.6.0/r2d2_sqlite/index.html#example) ?
Or [this](https://github.com/gwenn/r2d2-sqlite/blob/master/src/lib.rs#L25) if you need open flags.

**Comment by mqudsi** [comment]
Circling back, this is doable if you add a phantom generic type indicating whether or not the connection has a path, since that's explicitly a function of which API call was used to create the connection. (You can then actually move `path`  to a specialized function of the pathed connection, and get rid of the `Option` entirely. ) Then `Clone` can be implemented just for the specialized type with an explicit path.

However, I'm not sure if it should be implemented since this can result in multiple `DatabaseLocked` errors if multiple threads (or processes) simultaneously try merging their changes from a journal back to the database.

**Comment by fossdd** [comment]
Hey, looking back to it: It's still a nice idea.

**Comment by okpatil4u** [comment]
@mqudsi thank you very much for the commit !
How long do you think it will take until this feature is stable ?

**Comment by thomcc** [comment]
So, it hasn't been submitted as a PR, but I took a peek. Here's a preliminary review, assuming @mqudsi intends for a PR. Caveat: I'm somewhat against this, but not to the point where it's wontfix or anything.

---

So, first off, I think using `Clone` here is not the way we want to go. This should probably be `pub fn reopen(&self) -> Result<Connection>`. This will be correctly fallible, and while it won't allow people to `#[derive(Clone)]` or `conn.clone()`, it provides functionally the same thing. It also gives the right idea around what's happening — we're reopening the same connection.

---

I think we should outright forbid doing this for the `Connection::from_handle` connection. I'm worried that there's a decent amount of (unsafe) code that assumes this isn't possible, and while in principal we can audit it and change the safety requirement for `from_handle` to make this work... It seems quite tricky and error-prone (I don't really know everything that could cause problems, and without going over the SQLite docs closely at the same time as the unsafe code... It's hard for me to be sure).

As the most obvious (to me) example of what I mean, `Connection` is `Send`, so cloning a `Connection::from_handle`, and sending one to another thread would allow concurrent access to a Connection, which is UB, unless we configure SQLite in serialized mode (which we can't really rely on, and and even then, not all of SQLite's APIs are thread-safe in that mode).

That said, I expect in practice this isn't a problem — I doubt that anybody who does `Connection::from_handle` would even *want* this behavior, and if they do they can do `Connection::from_handle` twice (or use the API we have for getting the handle and do it that way), this would be unsafe, but that's reality.

---

There are two issues for the non-disc path connections:

1. In general, part of why apis like `std::fs::File` can't be cloned/reopened (and if they were, it would presumably be based on `dup*`/`DuplicateHandle`) is that the file you open might not be the same you initially opened. This does seem unlikely for a SQLite database, admittedly, but is a concern.

2. It's pretty likely to cause people to hit `SQLITE_BUSY` errors if they open multiple connections and write to more than one of them. This is a really hard to fix error, and plagues many applications that use SQLite in significant ways — the only real workaround I know is to only ever have one connection doing writes, and many doing reads, which I worry this encourages too much.

That said, naming the API `reopen` and documenting this it might be sufficient for the first of these not to be an issue, and the second I guess I could be convinced I'm being too opinionated, and shouldn't restrict valid uses because I think they're dubious and error-prone, unless the issues were well documented which they aren't, even in the official docs (and I'm no longer as familiar with why these things happen as I was when I worked on firefox's SQLite code...)

---

Finally, for in-memory DBs... I actually think this is nice. The concerns you mention are valid around SQLITE_LOCKED, although unlike SQLITE_BUSY (which IIRC shared-cache dbs don't get — they get SQLITE_LOCKED), it's largely fixed by unlock_notify. I'm not sure if `unlock_notify` should be on by default, but if we put this behind a feature, we could make that feature depend on `unlock_notify`.

Admittedly, it's a bit small to put behind a feature, so maybe that's dubious.

**Comment by mqudsi** [comment]
Wow, you guys are fast. I didn't intend to share that with anyone but was pushing the code to my fork for posterity. I'll address some of the feedback in all cases:

* There was never an intention to open a PR with this implemented as `Clone` due to the obvious fallibility issues. There *is* one way, but it would be involve delaying either the actual opening of the connection or the reporting of errors encountered in opening the connection to the first fallible operation post-clone, but that's just messy and not worth it to enable composite types to be able to derive `Clone` automatically. The commit uses `Clone` just as an initial PoC (mainly because I didn't have a better name) but I like the `reopen()` suggestion name and syntax.
* Once you accept that the operation may/must be fallible, then forbidding an unpathed connection opened via `from_handle()` to be reopened makes perfect sense. I agree that if you used the unsafe approach to get a db connection once, you're more than capable of doing it yourself again and don't need any help from the library.
* Regarding `:memory:` - I played around with a redesigned `Connection::path` (which would ideally be renamed) trying to get the semantics right. Right now, `Connection::path` is an `Option<PathBuf>` but the semantics of that aren't right since it's actually a connection string and not a file path, but it does make the 99% of the time when the connection string and the path are one and the same simpler. Technically, it should probably be something like `enum { Path(PathBuf), ConnectionString(String/OsString), Handle }` and we can validate that a `Path` passed into the `Connection` constructor is actually a path and exists on disk then reject a `reopen()` if the connection _wasn't_ opened with a real path, but I think the fact that the API already offers a function to create a new in-memory database (which would support reopening) makes it fair to just document the `reopen()` behavior as "btw, if you use `:memory:` it'll open a distinct db altogether - use the `open_in_memory(...)` api instead" and call it a day.

Now regarding `SQLITE_BUSY` and `SQLITE_LOCKED` - this is the real meat and potatoes. Enabling WAL makes a huge difference here (readers are no longer blocked by writers), and I would say there's definitely an argument to be made that without either WAL or read uncommitted, there's no point in a `reopen()` API, but I'm not sure how much leeway we have in terms of defaulting to WAL journaling? It's arguably an implementation detail of the db-access abstraction (rusqlite) and we could make it opt-out either by having users issue their desired journal mode as a pragma (opposite of what we do today to enable WAL) or by adding it as a flag when opening the database. After WAL, the obvious thing is to avoid any long-running write transactions that could lock out other readers/writers, but SQLite is pretty fast so if you have all the data you want to insert already/mostly prepared, the default 5s busy handler is more than enough (obviously some workloads differ) - but as you hinted, we don't have to settle there.

What I wanted to do is ideally gate the entire `reopen()` api behind `unlock_notify` and default to opening+reopening connections (in-memory or on-disk) with shared cache enabled (which only changes the locking semantics and reduces memory usage, but should not affect ACID in any way) unless `SQLITE_OPEN_PRIVATE_CACHE` was specified. Assuming process-exclusive access, that converts all SQLITE_BUSY situations to SQLITE_LOCKED because all connections behind a shared cache are treated as a single (multithreaded) connection, and then the SQLite unlock notification api would handle that better than the busy wait handler, and iirc, it does its own deadlock detection so we don't even need to worry about a timeout in the default case.

**Comment by thomcc** [comment]
> Wow, you guys are fast. I didn't intend to share that with anyone but was pushing the code to my fork for posterity

Ah, sorry. I would have ignored it except for someone asking about it. In the past there have been a lot of people using forks of rusqlite via `git` (largely because I've been slow merging stuff 😞... Still an issue) and I'd like to stay ahead of that where possible.

> but I'm not sure how much leeway we have in terms of defaulting to WAL journaling? It's arguably an implementation detail of the db-access abstraction (rusqlite)

I don't really think this is true... `rusqlite` isn't a generic data storage abstraction, it's a SQLite abstraction. We explicitly are not an ORM.

That is, we try relatively hard to do nothing more than do what direct SQLite usage should do, beyond providing safety, idiomatic abstractions, etc.

Many users are coming from other langs, or connecting to DBs which are not exclusively managed by `rusqlite`.... And more broadly, I tend to think it's not `rusqlite`'s place to choose different defaults than SQLite has chosen much [^most] of the time (and if we do, our choice should not cause problems for people, nor be too opinionated).

[^most]: There are, of course, a few exceptions, such as turning on multithreading support, and tweaks needed on certain platforms...

I have plans/goals here, but I've moved them to the end of this comment, since they're kind of off-topic. See the "Rambling about docs" section.

---

> default to opening+reopening connections (in-memory or on-disk) with shared cache enabled

I don't think this is a good idea. SQLite's official documentation explicitly discourages the use of shared cache: https://sqlite.org/c3ref/enable_shared_cache.html (emphasis mine)

> Shared cache is disabled by default. It is recommended that it stay that way. In other words, do not use this routine. This interface continues to be provided for historical compatibility, but its use is discouraged. _**Any use of shared cache is discouraged**_. If shared cache must be used, it is recommended that shared cache only be enabled for individual database connections using the sqlite3_open_v2() interface with the SQLITE_OPEN_SHAREDCACHE flag.

While we wouldn't be using that API, the emphasized text seems to be talking about the entire feature rather than that API in particular.

This actually means I'm a little concerned about doing it by default for in-memory connections too, but I think it's *probably* fine for us to make for them, so long as we document the behavior and the change.

---

## Rambling About Docs

*(This is basically off topic so I've moved it to the end — it also has an absurd density of long footnotes...)*

Anyway, I'm aware that this isn't ideal for many users, who have no familiarity with SQLite, nor interest in reading its docs, which are very large and refer to the C interface (and not all Rust programmers can read C).

My ideal solution here is better docs, probably a mdbook containing a guide to using `rusqlite` that discusses "best practices", tradeoffs, etc. I've intended to do this for a while, as I see people asking *a lot* of the same questions on discord, gitter, etc; stuff like:

1. How should I set up my database? (Shortish answer[^setup])
2. How should I use `rusqlite` across multiple threads? (Short answer[^threads])
3. How do I insert several records at once for bulk inserts/updates? (Short answer[^batch])
4. How do I use `rusqlite` in `async` code. (Shortish answer[^async])
5. ...

A lot of this stuff would also be nice to have example code for too. That said, my desire for more docs is not really related to this issue... but should help explain my philosophy for `rusqlite` — I'd prefer documenting the behavior rather than doing anything that would surprise people or cause issues for them that they wouldn't have if they had not used rusqlite.

[^setup]: There are a handful of `PRAGMA`s that people probably want to set — WAL mode is one of them (and if you do WAL mode, you should also set `PRAGMA synchronous` to `NORMAL`).

    That said, these are not good for *all* applications, just many, and use of WAL mode has some minor caveats (checkpointing, multiple files, etc) which generally don't cause problems but are worth mentioning, and can cause issues in some cases.

    So I don't think it's rusqlite's job to do it by default, and better handled by explaining the tradeoffs to the user, and letting them decide.

[^threads]: First: You should almost certainly turn on WAL mode. Then, the basic idea is that you want at most one writer connection (shared across all threads, if needed) and several reader connections. If you have multiple writer connections (even just two!), you'll probably get queries occasionally failing with SQLITE_BUSY 😞...

    While this sounds a bit like RwLock, it has better concurrency — in WAL mode, readers do not block writers and a writer does not block readers. (That said, readers can block checkpointing, so the `-wal` file can get large if you have very long-lived read transactions, which caused some issues in https://gitlab.com/famedly/conduit that I helped them with on Discord, and seems worth noting in docs).

    Anyway, there are a lot of ways to achieve this. A connection pool library can probably help (I'm not familiar with any of them though). Even just keeping a thread-local connection might be appropriate if you don't have a very large number of threads.

[^async]: Multiple options, none are great (at the end of the day, SQLite is a synchronous API, so there's not much we can do here without adding large overheads). Probably the best is `spawn_blocking`, or proxing messages back and forth to a worker thread (or pool), etc — the code to do this isn't too complex.

    `sqlx` is also a fine library and if you require async support and don't need some of the advanced SQLite features that it doesn't support, but note that it has meaningfully worse perf compared to synchronous usage (this isn't a dig — it's a good library but it just has to do more because SQLite is fundamentally synchronous).

    Given that `async` is a performance feature (which is unlike JS/Python/etc — for Rust, writing async code is substantially more complex than synchronous code, even if that synch. code has multiple threads), if SQLite is a core part of the application, it's worth questioning if use of `async` is right.

[^batch]: Actually there's no single short answer, specifics depend on the case.

    Most applications will be fine with beginning a transaction, preparing a statement that inserts/updates one row, executing it in a loop, and committing the transaction. In some cases building a query on the fly, or using the `array` feature are also options (although our `array` feature has some performance issues 😞). In rare cases, inserting into an in-memory connection that gets attached can be appropriate too...

    That said, the choice here doesn't matter as much as you might think in practice, but people get hung up on it a lot.

**Comment by gwenn** [comment]
> defaulting to WAL journaling

https://www.sqlite.org/wal.html
**WAL does not work over a network filesystem**
So, AFAIK, `rusqlite` cannot set it by default...

**Comment by thomcc** [comment]
Yeah, there are other reasons too. WAL is not *always* better, just usually.

**Comment by thomcc** [comment]
It might be worth asking on https://sqlite.org/forum/forum about shared cache mode? I agree it's a bit confusing that some docs say it's bad and others only talk about the benefits.

I wonder if its an idea that just didn't pan out well in practice...

**Comment by ekanna** [comment]
Async rusqlite built on top of rusqlite.
https://github.com/programatik29/tokio-rusqlite

**Comment by mqudsi** [comment]
> I wonder if its an idea that just didn't pan out well in practice...

Your gut was right on this one. [I finally got around to asking](https://sqlite.org/forum/forumpost/5489305f601b8c3f) on the SQLite forums for clarification on why the docs seem split over shared cache mode. I didn't receive any technical reasons for why we shouldn't use shared cache mode, but did get an official (loud and clear) answer that use of shared cache mode is discouraged. The SQLite docs have also been updated in response to my query.

This doesn't necessarily put the kibosh on implementing `Clone` but it does change the calculus and it means that we'll have to lean heavily into documenting the best practices for using SQLite from multiple threads rather than merely working around those limitations with shared cache mode and unlock-notify.

**Comment by bubelov** [comment]
@thomcc can you clarify what do you mean here?

> You should almost certainly turn on WAL mode. Then, the basic idea is that you want at most one writer connection (shared across all threads, if needed) and several reader connections. If you have multiple writer connections (even just two!), you'll probably get queries occasionally failing with SQLITE_BUSY disappointed...

I'm thinking on a good way to integrate `rusqlite` into a modern async webserver and here is my current understanding:

- `spawn_blocking` is too verbose and ignoring an async runtime is probably not a big deal, I expect it to behave like a thread pool based webserver where one of the threads should block till it handles a request. There might be some overhead in treating an async runtime as a dumb thread pool but I didn't find any data on that so far.
- WAL mode is must have since it allows for better concurrency.
- Giving each thread a separate db connection should improve read performance.
- Write performance doesn't improve with the number of threads and/or CPUs.

That led me to consider giving each thread a separate SQLite connection and allowing those connections to perform both reads and writes, letting SQLite to manage locking. I don't really expect to encounter `SQLITE_BUSY`, except for the edge cases listed in the doc:

1. If another database connection has the database mode open in [exclusive locking mode](https://www.sqlite.org/pragma.html#pragma_locking_mode) then all queries against the database will return [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy). Both Chrome and Firefox open their database files in exclusive locking mode, so attempts to read Chrome or Firefox databases while the applications are running will run into this problem, for example.

2. When the last connection to a particular database is closing, that connection will acquire an exclusive lock for a short time while it cleans up the WAL and shared-memory files. If a second database tries to open and query the database while the first connection is still in the middle of its cleanup process, the second connection might get an [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) error.

3. If the last connection to a database crashed, then the first new connection to open the database will start a recovery process. An exclusive lock is held during recovery. So if a third database connection tries to jump in and query while the second connection is running recovery, the third connection will get an [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) error.

Regarding the first warning, the default locking mode is `NORMAL`, so I don't think that concurrent writes should trigger `SQLITE_BUSY`.

Am I missing something?
